python/ray/data/_internal/logical/interfaces/plan.py

```diff
@@ -8,6 +13,13 @@ class Plan:
     that's useful for plan optimization or execution.
     """
 
+    def __init__(self):
+        self._context: Optional["DataContext"] = None
+
     @property
     def dag(self) -> Operator:
         raise NotImplementedError
+
+    def _set_context(self, context: Optional["DataContext"]):
+        """Set the DataContext for this plan, usually from its ExecutionPlan."""
+        self._context = context
```

Comment:
Curious to get @raulchen's thoughts too, but I'm wondering if we should make this a parameter in the `Plan` constructor? Think it'd improve readability (by making it clear that a context must be provided) and make the object more immutable (by ensuring that the context is set exactly once)

Reply:
the main reason i didn't want to make it an explicit parameter for `Plan` is because we sometimes haven't constructed the `DataContext` or an `ExecutionPlan` containing a sealed `DataContext` at the time of creating a `LogicalPlan`. For example in [many read_api](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L399-L413)s, we construct the logical operator and logical plan, then construct a `ExecutionPlan` object separately afterwards.

Reply:
> although, i suppose it's possible to construct the ExecutionPlan first before creating the logical plan.

This seems a bit cleaner. I slightly prefer this if not requiring too many changes.

Reply:
yeah no problem, let me refactor it.